### PR TITLE
Irsa 2146 add naifid resolver to firefly

### PR DESCRIPTION
**Pull request by deannaji**
Kindly review:
This ticket is about adding a Naifid Panel to firefly.

Jira: https://jira.ipac.caltech.edu/browse/IRSA-2146
K8s: https://irsawebdev9.ipac.caltech.edu/irsa-2146/firefly/

Note: For testing, please: 
- Go to (https://irsawebdev9.ipac.caltech.edu/irsa-2146/firefly//firefly-dev.html)
- Click on 'Test Searches' button (on the top menu), then click on the 'NAIF-ID' tab (the last tab). 
- Or, click on 'Examples Js Dialog' button.

### CODE REVIEW

File: src/firefly/java/edu/caltech/ipac/firefly/server/rpc/ResolveServerCommands.java [hunk 1]
```diff
@@ -37,7 +41,40 @@ public String doCommand(SrvParam sp) throws Exception {
             }
         }
 
-        public boolean getCanCreateJson() { return false; }
+        public boolean getCanCreateJson() { return true; }
```
Comment (loitly) [inline src/firefly/java/edu/caltech/ipac/firefly/server/rpc/ResolveServerCommands.java]: This is why TargetPanel stopped working.

File: src/firefly/js/ui/NaifidPanel.jsx [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+            const rval = resolveNaifidObj(val);
+            if (!rval.p) return undefined;
+            return rval.p.then((response)=>{
+                if(response.valid) {
+                    let suggestionsList = response.data || [];
+
+                    if(this.iAmMounted && this.state.suggestions !== suggestionsList) {
+                        this.setState({suggestions: suggestionsList});
+                    }else if(this.state.suggestions === suggestionsList){
+                        this.setState({suggestions:''});
```
Comment (loitly) [inline src/firefly/js/ui/NaifidPanel.jsx]: If you're checking for `this.iAmMounted` before `setState` on top, I am pretty sure you need to  check here too.

File: src/firefly/js/ui/NaifidPanel.jsx [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+        if (get(t,'worldPt')) retVal= t.worldPt.toString();
+    }
+    return retVal;
+}
+
+function getNaifidValue(onCallBack) {
+    return (val, str) => {
+        if (! str) return;
+
+        const naifId = (str.match(/Naifid:.*/))[0].split('Naifid:').pop();
```
Comment (loitly) [inline src/firefly/js/ui/NaifidPanel.jsx]: This is error-prone.  When match fail, `[0]` will cause TypeError.
Should replace with this:
`const [,naifId] = str.match(/Naifid:(.*)/) || [];`

File: src/firefly/js/ui/NaifidPanelWorker.js [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    if (!objName) {
+        return {
+            showHelp: true,
+            valid : true,
+            feedback: ''
+        };
+    }
+
+    let {p}= makeResolverPromise(objName);
+    p= p.then( (results) =>
```
Comment (loitly) [inline src/firefly/js/ui/NaifidPanelWorker.js]: This is misleading... why re-assign `p` to a `then` declaration?
It should be written,  `p.then( (results) =>`  without `p = `.

Comment (robyww) [reply src/firefly/js/ui/NaifidPanelWorker.js]: OK, the problem is that you want to return the then promise not p. So what @deannaji is doing is correct, but confusing. It should just be `return p.then...`

But if you do then you will needs to modify `getSuggestions` so it is expect a `promise` from `resolveNaifidObj` and not an `object`.
 
NaifidPanel.jsx, `getSuggestions`
```
            if (!rval) return undefined;
            return rval.then((response)=>{
```

I think you copied some code from target panel where it was return not only a promise in the object but also a function to abort the promise. Since you are only returning a promise there is no need to wrap it in an object.

File: src/firefly/java/edu/caltech/ipac/firefly/server/rpc/ResolveServerCommands.java (hunk 1 above)
Comment (deannaji) [reply src/firefly/java/edu/caltech/ipac/firefly/server/rpc/ResolveServerCommands.java]: Done, thanks!

**Comment by robyww** [comment]
> The dropdown menu should be above the form and height should be larger to show 5 items at least (currently i see only 2)

This is a problem with the disclosure panel. It is getting cut off by the wrapping component.  This is fixable but not a big problem because almost all panels are bigger and therefore it is not cut off. It should be a bug in a separate ticket.

**Comment by robyww** [comment]
> Dropdown label should be in one line: 'Name: .. , Naifid: …', again, this should be controlled by a property css

Tested this on the most recent push, looks good

File: src/firefly/js/ui/NaifidPanel.jsx (hunk 2 above)
Comment (deannaji) [reply src/firefly/js/ui/NaifidPanel.jsx]: Done, thanks!

File: src/firefly/js/ui/NaifidPanel.jsx (hunk 3 above)
Comment (deannaji) [reply src/firefly/js/ui/NaifidPanel.jsx]: Done!
Thanks :)

File: src/firefly/js/ui/NaifidPanelWorker.js (hunk 4 above)
Comment (deannaji) [reply src/firefly/js/ui/NaifidPanelWorker.js]: Done, thanks!!

**Comment by deannaji** [comment]
I did update K8s build, and done very minor tweaks to make it look like what we see here:
(https://sha.ipac.caltech.edu/applications/Spitzer/SHA/#id=MOSQuery&RequestClass=ServerRequest&shortDesc=Solar%20System%20Object/Orbit%20Precovery%20Search&isBookmarkAble=true&isDrillDownRoot=true)

Thanks for help!
:)

File: src/firefly/js/ui/NaifidPanelWorker.js [hunk 5]
```diff
@@ -93,7 +93,7 @@ function resolveObject(objName) {
     }
 
     let {p}= makeResolverPromise(objName);
-    p= p.then( (results) =>
+    p.then( (results) =>
```
Comment (loitly) [inline src/firefly/js/ui/NaifidPanelWorker.js]: My mistake.  Trey corrected me.  You had it right before.
It was just awkward to see it written that way.  Please revert this change.

Comment (deannaji) [reply src/firefly/js/ui/NaifidPanelWorker.js]: Completed, Thanks!!

**Comment by deannaji** [comment]
@ejoliet , can I merge and fix other issues in a different ticket. So I can add NAIFID to SOFIA :)

**Comment by deannaji** [comment]
Copy that, thanks!
